### PR TITLE
Limit LIMITED_RETRY

### LINKED ISSUE #1706: Don't wait forever for servers to transition out of the `WAIT` destiny
From @manishtomar's review on #1702:

> I am worried about convergence running forever due to servers remaining in Destiny.WAIT forever. Honestly, I do not trust Nova to ensure servers will become normal. I think we should put the group in ERROR if Destiny.WAIT servers do not become ACTIVE within a certain time. It can be done in another PR though :)

I think the logic should be something like: if the only thing we're doing is waiting for servers in WAIT, and we do this for more than a certain period of time (either in terms of convergence iterations or real-time), stop iterating and put the group into ERROR.

### PR DESCRIPTION
**Pull request by radix**
This branch fixes #1706 the rest of the way.

I tore through the tests a bit more refactoring them.

### CODE REVIEW

**Comment by radix** [comment]
Jenkins please run integration tests

**Comment by radix** [comment]
Jenkins please run integration tests

**Comment by radix** [comment]
Jenkins please run integration tests

File: otter/convergence/service.py [hunk 1]
```diff
@@ -510,6 +511,10 @@ def converge_one_group(currently_converging, recently_converged,
     :param callable execute_convergence: like :func`execute_convergence`, to
         be used for test injection only
     """
+    clean_waiting = waiting.modify(
+        lambda group_iterations:
+            group_iterations.remove(group_id)
```
Comment (manishtomar) [inline otter/convergence/service.py]: How about using [`group_iterations.discard`](http://pyrsistent.readthedocs.org/en/latest/api.html#pyrsistent.PMap.discard)?

File: otter/convergence/service.py [hunk 2]
```diff
@@ -531,6 +537,20 @@ def converge_one_group(currently_converging, recently_converged,
         # unexpected errors, so convergence will be retried.
         yield err(None, 'converge-non-fatal-error')
     else:
+        if result[0] is StepResult.LIMITED_RETRY:
+            # We allow further iterations to proceed as long as we haven't been
+            # waiting for a LIMITED_RETRY for 10 consecutive iterations.
+            if (yield waiting.read())[group_id] > 10:
```
Comment (manishtomar) [inline otter/convergence/service.py]: Can 10 be configurable value? Unfortunately, it might have to become another argument :expressionless:

Comment (radix) [inline otter/convergence/service.py]: Oh yes, I meant to and forgot to add it. thanks :)

File: otter/convergence/service.py (hunk 1 above)
Comment (radix) [inline otter/convergence/service.py]: Ah, I always forget about that method!

**Comment by radix** [comment]
jenkins please run integration tests

File: otter/convergence/service.py [hunk 3]
```diff
@@ -531,6 +539,20 @@ def converge_one_group(currently_converging, recently_converged,
         # unexpected errors, so convergence will be retried.
         yield err(None, 'converge-non-fatal-error')
     else:
+        if result[0] is StepResult.LIMITED_RETRY:
+            # We allow further iterations to proceed as long as we haven't been
+            # waiting for a LIMITED_RETRY for N consecutive iterations.
+            if (yield waiting.read())[group_id] > limited_retry_iterations:
+                yield msg('converge-limited-retry-too-long')
```
Comment (manishtomar) [inline otter/convergence/service.py]: Entry in `log/spec.py`? :)

File: otter/convergence/service.py [hunk 4]
```diff
@@ -531,6 +539,20 @@ def converge_one_group(currently_converging, recently_converged,
         # unexpected errors, so convergence will be retried.
         yield err(None, 'converge-non-fatal-error')
     else:
+        if result[0] is StepResult.LIMITED_RETRY:
+            # We allow further iterations to proceed as long as we haven't been
+            # waiting for a LIMITED_RETRY for N consecutive iterations.
+            if (yield waiting.read())[group_id] > limited_retry_iterations:
```
Comment (manishtomar) [inline otter/convergence/service.py]: Correct me if I am wrong but wouldn't it raise `KeyError` when group returns LIMITED_RETRY the first time? Maybe you need `defaultdict` here?

Comment (radix) [inline otter/convergence/service.py]: oh yes! I briefly had the thought to make sure I was handling this correctly but I forgot. Thanks very much for spotting it!

**Comment by radix** [comment]
jenkins please run integration tests

File: otter/convergence/service.py [hunk 5]
```diff
@@ -531,6 +539,20 @@ def converge_one_group(currently_converging, recently_converged,
...
         yield err(None, 'converge-non-fatal-error')
     else:
+        if result[0] is StepResult.LIMITED_RETRY:
+            # We allow further iterations to proceed as long as we haven't been
+            # waiting for a LIMITED_RETRY for N consecutive iterations.
+            current_iterations = (yield waiting.read()).get(group_id, 0)
+            if current_iterations > limited_retry_iterations:
+                yield msg('converge-limited-retry-too-long')
+                yield clean_waiting
+                yield delete_divergent_flag(tenant_id, group_id, version)
```
Comment (manishtomar) [inline otter/convergence/service.py]: The group is NOT put in ERROR in this case. It is done inside `execute_convergence`. Now, this makes me wonder about the boundary between `execute_convergence` and `converge_one_group`.

Comment (radix) [inline otter/convergence/service.py]: sorry, you're right, I'll fix that. Yeah, the boundary between them stinks, as does the factoring of the whole module :(

Comment (manishtomar) [inline otter/convergence/service.py]: No problem :) Sorry I didn't intend to be rude. Just wanted to be clear :)

**Comment by radix** [comment]
jenkins please run integration tests

**Comment by radix** [comment]
okay, so I completely redid this so that execute_convergence does the work and also puts the group into ERROR as appropriate.

**Comment by radix** [comment]
jenkins please run integration tests

**Comment by radix** [comment]
There seems to be a real reproducible bug with this PR -- I'll try to debug

**Comment by radix** [comment]
I found the bug here. Created #1743 to help address it; changes to this PR incoming soon.

**Comment by radix** [comment]
YES! Got those failing tests to pass on my VM :D This PR is still in-progress while waiting on #1743 and #1742 (which are currently already merged into this branch)

**Comment by radix** [comment]
jenkins please run integration tests

File: otter/convergence/service.py [hunk 6]
```diff
@@ -283,12 +293,29 @@ def execute_convergence(tenant_id, group_id, build_timeout,
               desired=desired_group_state)
     worst_status, reasons = yield _execute_steps(steps)
 
+    if worst_status != StepResult.LIMITED_RETRY:
+        # If we're not waiting any more, keeping track of the group
```
Comment (manishtomar) [inline otter/convergence/service.py]: I guess you meant "If we're not waiting any more, **there is no point** keeping track of the group"

File: otter/convergence/errors.py [hunk 7]
```diff
@@ -86,3 +95,6 @@ def String(string):
         return {'string': string}
 
     Structured = identity
+
+    def UserMessage(message):
```
Comment (manishtomar) [inline otter/convergence/errors.py]: Any reasons for `UserMessage`?

**Comment by manishtomar** [comment]
LGTM modulo nitpick above :)

File: otter/convergence/errors.py (hunk 7 above)
Comment (radix) [inline otter/convergence/errors.py]: It basically just gives us a convenient way to specify a user-presentable message from ad-hoc planning code. I use it in the change in planning.py.

**Comment by radix** [comment]
Thank you @manishtomar! :)

**Pull request merged**